src/txmempool.cpp

```diff
@@ -120,47 +120,91 @@ void CTxMemPool::UpdateForDescendants(txiter updateIt, cacheMap& cachedDescendan
                                       const std::set<uint256>& setExclude, std::set<uint256>& descendants_to_remove,
                                       uint64_t ancestor_size_limit, uint64_t ancestor_count_limit)
 {
-    CTxMemPoolEntry::Children stageEntries, descendants;
-    stageEntries = updateIt->GetMemPoolChildrenConst();
-
-    while (!stageEntries.empty()) {
-        const CTxMemPoolEntry& descendant = *stageEntries.begin();
-        descendants.insert(descendant);
-        stageEntries.erase(descendant);
-        const CTxMemPoolEntry::Children& children = descendant.GetMemPoolChildrenConst();
-        for (const CTxMemPoolEntry& childEntry : children) {
-            cacheMap::iterator cacheIt = cachedDescendants.find(mapTx.iterator_to(childEntry));
-            if (cacheIt != cachedDescendants.end()) {
-                // We've already calculated this one, just add the entries for this set
-                // but don't traverse again.
-                for (txiter cacheEntry : cacheIt->second) {
-                    descendants.insert(*cacheEntry);
+    const CTxMemPoolEntry::Children& children = updateIt->GetMemPoolChildrenConst();
+
+    // initialize to hold all direct children
+    // children guaranteed to be unique at this point
+    std::vector<txiter> descendants(children.size());
+    size_t i = 0;
+    for (const auto& child: children) {
+        descendants[i] = mapTx.iterator_to(child);
+        ++i;
+    }
+
+    {
+        WITH_FRESH_EPOCH(m_epoch);
+        // visit all children
+        for (auto& child : descendants) {
+            visited(child);
+        }
+        for (size_t i = 0, n_to_process = descendants.size(); i < n_to_process; ++i) {
+            const CTxMemPoolEntry& descendant = *descendants[i];
+            const CTxMemPoolEntry::Children& children = descendant.GetMemPoolChildrenConst();
+            for (const CTxMemPoolEntry& childEntry : children) {
+                cacheMap::iterator cacheIt = cachedDescendants.find(mapTx.iterator_to(childEntry));
```

Comment:
i think you are right on this. The patch here is a 100% behavioral match, but it does seem like one could evade caching a little because we don't check if children are cached. Can you think of any functional differences? (this code is incredibly subtle, so i'm reticent to change it).

Reply:
I think it should only be done if `children` is empty - this change should also allow getting rid of the current `find` call in the children loop and just add straight to `descendants`. The current logic in master of checking `cachedDescendants` is a bit odd as if there are three txn's (t1 -> t2 -> t3) in a chain that depend on each other sequentially, then the `find` call will only get a "hit" when t3 is `updateIt` as it "skips" an entry